src/store/filestore.rs

```diff
@@ -1,213 +1,216 @@
-        Self::parse_annotation_path(&self.make_annotation_path("pod", "*", name, "*"))?
-            .map(|metadata| -> Result<(String, String), Box<dyn Error>> {
-                let resolved_metadata = metadata?;
-                let hash = resolved_metadata.1 .0;
-                let version = resolved_metadata.1 .1;
-                Ok((version, hash))
-            })
-            .collect::<Result<BTreeMap<String, String>, _>>()
-    }
-
-    fn save_file(file: &Path, content: &str, fail_if_exists: bool) -> Result<(), Box<dyn Error>> {
-        fs::create_dir_all(file.parent().ok_or_else(|| FileHasNoParent {
-            path: file.to_path_buf(),
-        })?)?;
-        let file_exists = file.exists();
-        if file_exists && fail_if_exists {
-            return Err(Box::new(FileExists {
-                path: file.to_path_buf(),
-            }));
-        } else if file_exists {
+        }
+
+        Ok(())
+    }
+
+    /// Generic function for loading spec.yaml into memory
+    fn load_item<T: DeserializeOwned>(&self, item_key: &ItemKey) -> Result<T> {
+        match item_key {
+            ItemKey::NameVer(name, version) => {
+                // Search the name-ver index
+                let hash = self.get_hash_from_name_ver_tree::<T>(name, version)?;
+
+                // Get the spec and annotation yaml
+                let spec_yaml = fs::read_to_string(self.make_path::<T>(&hash, "spec.yaml"))?;
+
+                let annotation_yaml =
+                    fs::read_to_string(self.make_annotation_path::<T>(&hash, name, version))?;
+
+                from_yaml::<T>(&spec_yaml, &hash, Some(&annotation_yaml))
+            }
+            ItemKey::Hash(hash) => {
+                // Get the spec and annotation yaml
+                let spec_yaml = fs::read_to_string(self.make_path::<T>(hash, "spec.yaml"))?;
+                from_yaml::<T>(&spec_yaml, hash, None)
+            }
+        }
+    }
+
+    fn delete_item<T>(&self, item_key: &ItemKey) -> Result<()> {
+        let hash = match item_key {
+            ItemKey::NameVer(name, version) => {
+                // Search the name-ver index
+                self.get_hash_from_name_ver_tree::<T>(name, version)?
+            }
+            ItemKey::Hash(hash) => hash.to_owned(),
+        };
+
+        fs::remove_dir_all(self.make_dir_path::<T>(&hash))?;
+        Ok(())
+    }
+
+    fn build_name_ver_tree<T>(&self) -> Result<BTreeMap<String, String>> {
+        // Construct the cache with glob and regex
+        let type_name = get_type_name::<T>();
+        let re = Regex::new(&format!(
+            r"^.*\/{type_name}\/(?<hash>[a-z0-9]+)\/annotations\/(?<name>[A-z0-9\- ]+)-(?<ver>[0-9]+.[0-9]+.[0-9]+).yaml$"
+        ))?;
+
+        // Create the missing cache btree for the item
+        let mut name_ver_tree = BTreeMap::new();
+
+        let search_pattern = self.make_dir_path::<T>("*").join("annotations/*");
+
+        for path in glob::glob(&search_pattern.to_string_lossy())? {
```

Comment:
In general, it is *usually* better to perform an iterator (lazy) approach as opposed to explicit loops. Here are a few reference articles for Rust specifically that can give some context where I am coming from.

- https://www.lurklurk.org/effective-rust/iterators.html
- https://doc.rust-lang.org/book/ch13-04-performance.html

Reply:
I prefer readability in this case since I didn't want to redefine the regex pattern every loop like you had in your previous example. Plus it had that odd return type.

Reply:
Let's put this as a point for a quick discussion

Reply:
@guzman-raphael Will look into a few cases and report back.